### DISCUSSION TITLE
How to create a cone shape ?

### DISCUSSION THREAD

**Discussion by finscn** [jrouwe/JoltPhysics#714] [General]
How to create a cone shape ?

In some engine , the cone is a special Cylinder , there are 2 radius , when one is 0 , the Cylinder will become a cone.
But I can't do that in JoltPhysics.

**Comment by jrouwe** [comment]
The best way to do it is to create a convex hull shape. You'd generate points on a circle (note: don't use too many, 8 or 16 is probably enough) and a single point for the top.

**Reply by finscn** [reply]
Could you supply a ConeShape directly  or make Cylinder support 2 radiuses that be like TaperedCapsule?
I think ConeShape is very common object, event more useful than TaperedCapsule.

**Reply by jrouwe** [reply]
Just like a cylinder shape, cone shapes are very nasty in simulation because they have edges that are both curved and sharp, this combination results in collision detection / contact manifold calculations being less accurate and often results in jittery simulation. 

Tapered capsules don't have this problem because they only have curved surfaces. Tapered capsules are very useful for limbs, so are actually one of the most commonly used shape.

While it would be possible to create a ConeShape, I think you'll get a more stable simulation using a ConvexHullShape and performance wise I don't think it would matter much either.

For reference: PhysX doesn't support cone shapes either (and it also doesn't support cylinder shapes).

**Reply by Pierre-Terdiman** [reply]
**_For reference: PhysX doesn't support cone shapes either (and it also doesn't support cylinder shapes)._**

That is not correct anymore. With the addition of PxCustomGeometry, it does support both.

See e.g. https://github.com/NVIDIA-Omniverse/PhysX/blob/main/physx/snippets/snippetcustomconvex/CustomConvex.h

But even in PEEL I used "real" cylinders in dedicated test scenes.

**Reply by jrouwe** [reply]
Ok, I stand corrected 😅

**Comment by venilark** [comment]
No need to open a new discussion, and what about a Torus?
for ring-like shapes or explosions like in old games would be helpful

**Reply by jrouwe** [reply]
A torus is not convex so you need to create a StaticCompoundShape with ConvexHullShape segments to represent the torus.

**Comment by aaronfranke** [comment]
This is now possible as of this PR: https://github.com/jrouwe/JoltPhysics/pull/1241
